feat(compile): Enable multiple roots for a standalone module graph

andreubotella opened the pull request:
This change will enable dynamic imports and web workers to use modules not reachable from the main module, by passing a list of extra side module roots as options to `deno compile`.

------

This PR builds on top of #17657, which adds support for web workers to `deno compile`.

The option name `--extra-side-modules` is up for bikeshedding, as well as the exact way to specify such list of modules. I suspect this list could be added to `deno.json`.

Closes #17908

dsherret commented on cli/tests/testdata/compile/dynamic_imports/main.js:
```diff
@@ -0,0 +1,6 @@
+
+setTimeout(() => {
+  console.log("Dynamic importing");
+  import("./import1.js").then(() => console.log("Dynamic import done."));
```
This is statically analyzable by deno_graph and should already be in the module graph. I don't think we need to require people to specify a CLI flag for this and I think it should just be done automatically?

andreubotella replied:
It is not currently being added to the module graph, and removing the CLI flag makes the test fail. But even if we can detect uses of dynamic import (and the `Worker` constructor, because of #17657) with string arguments, a CLI flag would still be useful for cases that can't be detected.

aapoalas replied:
Should we open an issue about deno_graph not apparently including statically analysable dynamic imports into the module graph?

dsherret replied:
It's there. We just have to make it work.

dsherret replied:
I investigated this and it's an eszip thing. Opened https://github.com/denoland/eszip/pull/115

Once that's merged then the dynamic import test that's statically analyzable doesn't need the flag, so let's wait for that to merge and I'll do an eszip release then we can bump eszip in this PR.

aapoalas commented on cli/args/flags.rs:
```diff
@@ -846,6 +847,13 @@ fn compile_subcommand<'a>() -> Command<'a> {
+      Arg::new("extra-side-modules")
+        .long("extra-side-modules")
+        .takes_value(true)
+        .multiple_occurrences(true)
```
**Bikeshedding:** Since this is multiple occurrence, should it be `--extra-side-module` instead? Plural in name to me suggests a comma separated list.

Also, maybe just `--side-module`? Or something like `--dynamic-module-entry-point`  :D

andreubotella replied:
I don't really care about the option name, and as I said in the OP, this could very well be an option in `deno.json` instead, or something else. I'll just let the Deno core team (and/or any contractors 😉) pick the name.

dsherret replied:
Since `deno compile` requires an entrypoint (and therefore, could apply to multiple files within a project), then it's probably best for it to be specified as a deno task. So a flag would be most appropriate. I think dropping the word "extra" for brevity and making it `--side-module` sounds good.

aapoalas commented on cli/tests/testdata/compile/workers/not_in_module_map.js:
```diff
@@ -1,7 +1,11 @@
 // This time ./worker.ts is not in the module map, so the worker
-// initialization will fail.
+// initialization will fail unless worker.js is passed as an extraside module.
```
"extraside"

aapoalas reviewed (approved):
Minor nitpicks and bikesheddings aside, looks good to me and really useful.

Statically analysable dynamic imports being picked up into the module graph automatically would be a really good feature to get working, as it would make both `deno compile` and Deno Deploy users quite happy indeed. For Workers I don't think it would work though, since it's possible to override the `Worker` name can be overridden both by local shadowing and globally by overriding `globalThis.Worker`.

dsherret reviewed (changes requested):
We should be able to do it without needing `--extra-side-modules`.

Edit: let's wait on https://github.com/denoland/eszip/pull/115

andreubotella commented:
> We should be able to do it without needing `--extra-side-modules`.

Some flag would still be needed for non-analyzable dynamic imports, and for workers.

dsherret commented:
Sorry, I read too fast and missed about workers. I think that's reasonable for workers and in that case maybe we can just merge this without the analyzing of statically analyzable dynamic imports. For some reason, all our walking/traversing of the graph in the CLI skips over statically analyzable dynamic imports, but they're in the `ModuleGraph`. I'll review this now.

andreubotella commented:
> Sorry, I read too fast and missed about workers. I think that's reasonable for workers and in that case maybe we can just merge this without the analyzing of statically analyzable dynamic imports. For some reason, all our walking/traversing of the graph in the CLI skips over statically analyzable dynamic imports, but they're in the `ModuleGraph`. I'll review this now.

Thanks. You should review commit 4d8e192ea8a881efb3a51ec82d6bf58587c43e2b, since the previous ones come from #17657. I built this PR off of that one not because the actual code changes need to be based on that PR, but because that way I can add to one of the tests from that PR.

> I agree with @andreubotella that the flag would be useful for non-analysable imports but... It's actually also possible to do that without the flag:
> 
> ```ts
> const EXTRA_SIDE_MODULES = [
>   () => import("./side-module-1.ts"),
>   () => import("../root-side-module.ts"),
> ];
> ```
> 
> It's probably a matter of taste which is more elegant: Use the existing language features that we have, or add a configuration flag or option.

This seems like a hard-to-explain non-obvious hack. It feels a lot cleaner to me to say "`deno compile` will try to detect dynamic imports, but it can't always, so use this flag just in case".

nayeemrmn commented:
> For some reason, all our walking/traversing of the graph in the CLI skips over statically analyzable dynamic imports

Not sure if you're talking about somewhere specific, but the logic of skipping over (even statically analyzable) dynamic imports exists for example in startup graph validation because we don't want to elevate dynamic module slot errors until the runtime import of that module happens.

dsherret commented:
> Not sure if you're talking about somewhere specific, but the logic of skipping over (even statically analyzable) dynamic imports exists for example in startup graph validation because we don't want to elevate dynamic module slot errors until the runtime import of that module happens.

Yeah, why is that done? I was thinking that maybe someone would be downloading code on the fly, but that wouldn't work. It seems to shift an error that could be spotted at startup to being surfaced later on. Anyway, in this case we can make it work by updating eszip to not skip over dynamic imports: https://github.com/denoland/eszip/pull/115

nayeemrmn commented:
> Yeah, why is that done? I was thinking that maybe someone would be downloading code on the fly, but that wouldn't work. It seems to shift an error that could be spotted at startup to being surfaced later on.

Because the import might not happen, and even be conditioned on whether or not it would error:
```ts
if (fs.exists("plugin.ts")) {
  await import("./plugin.ts")
}
```

dsherret commented:
@nayeemrmn that makes sense. Thanks for the background!

nayeemrmn commented:
Re flag name, the term 'side module' in this context seems too colloquial doesn't it? `--imports`, `--include` or even just `--modules` would be better IMO.

aapoalas commented:
> Re flag name, the term 'side module' in this context seems too colloquial doesn't it? `--imports`, `--include` or even just `--modules` would be better IMO.

I would vote for `--import` or `--include`.

aapoalas commented:
Potentially related and weird errors in in WPT tests for workers:

```
/html/webappapis/atob/base64.any.worker.html


file result: ok. 380 passed; 0 failed; 0 expected failure; total 380

error: Uncaught (in worker "") TypeError: globalThis_.addEventListener is not a function
    at promiseRejectMacrotaskCallback (internal:cli/runtime/js/99_main.js:376:17)
----------------------------------------
/webidl/ecmascript-binding/global-immutable-prototype.any.worker.html

test stderr:
error: Uncaught (in worker "") TypeError: globalThis_.addEventListener is not a function
    at promiseRejectMacrotaskCallback (internal:cli/runtime/js/99_main.js:376:17)

file result: failed. runner failed during test
```

First one doesn't fail but it's a bit suspicious. Coming from the 99_main.js file:
```js
    // Add a callback for "error" event - it will be dispatched
    // if error is thrown during dispatch of "unhandledrejection"
    // event.
    globalThis_.addEventListener("error", errorEventCb);
```

dsherret commented on cli/tests/integration/compile_tests.rs:
```diff
@@ -564,3 +564,118 @@ fn check_local_by_default2() {
+  let output = util::deno_cmd()
+    .current_dir(util::root_path())
+    .arg("compile")
+    .arg("--side-module")
+    .arg(util::testdata_path().join("./compile/dynamic_imports/import1.ts"))
```
If you bump eszip to 0.36.0 (just released) then you should be able to remove these two lines.

andreubotella replied:
I was planning on doing that, and then add another test with a dynamic import that isn't easily analyzable. Maybe something where the relative URL is read from a file on disk, so the test is useful even if the static analysis improves.

dsherret commented on cli/args/flags.rs:
```diff
@@ -846,6 +847,13 @@ fn compile_subcommand<'a>() -> Command<'a> {
     .trailing_var_arg(true)
     .arg(script_arg().required(true))
+    .arg(
+      Arg::new("side-module")
```
It would be good to add some help text for what this argument does.

piscisaureus replied:
I don't think it's appropriate to make this a command line argument. The fact that a module requires certain modules to function is information about the program that is 100% determined by the author and not by the user. So it should be specified by the author, maybe in package.json or a pragma or... they could add `#import "that-module";` somewhere else in their code.

andreubotella replied:
@aapoalas suggested something similar in https://github.com/denoland/deno/pull/17663#issuecomment-1426832462. I have since realized that unreachable dynamic imports added somewhere on the main thread's module map wouldn't really work for web workers when typechecking is on.

It seems like when a module is added as a root of the module graph, with this command line argument, it is by default typechecked with the `"deno.window"` lib, but if it includes a triple-slash directive, it will use the directive's lib. That is not ideal, but it's workable. However, when a module is reachable from the main thread's module graph, it will use the `"deno.window"` lib _regardless of any triple-slash directives_. While I haven't tested the static analysis in the new version of eszip yet, I suspect it will work like that as well.

piscisaureus replied:
> However, when a module is reachable from the main thread's module graph, it will use the "deno.window" lib in addition to any triple-slash directives, resulting in duplicate type definitions. While I haven't tested the static analysis in the new version of eszip yet, I suspect it will work like that as well. Meaning that this "hack" wouldn't really work for web workers.

That seems like a detail (potentially a bug?) that can be worked out or fixed. Type check them both with and witout default lib maybe?

piscisaureus replied:
Maybe a bare import statement could be done from the main module that runs in the worker? If deno spots it there it can assume it is run from a worker.

aapoalas replied:
There is no "main module that runs in the worker" since the `Worker` constructor isn't statically analysable (it can be overridden both locally and globally). This PR is essentially about providing that main worker module.

Not an easy problem to solve. On one hand you'd expect dynamic imports to be type checked, on the other hand you want to avoid them because they're probably made dynamic for a reason, f.ex. being platform specific. For the hack that I proposed, adding modules to the module graph for having them be available for Worker scripts, it's beneficial that they're not type checked but also their inclusion into the module graph must not place any type libs (`"deno.window"`) on them either...

The real answer would be the [asset reference proposal](https://github.com/tc39/proposal-asset-references) though even then the question of how to do types is a bit tough.

piscisaureus replied:
I'd suggest to treat these as orthogonal problems: the fact that it is not possible for a module to declare a dependency on a worker script is a different problem that this PR doesn't need to solve. 

It is also a real problem though esp. when a "complete" eszip must be built (e.g. when deploying to deploy).

dsherret replied:
Maybe we could have deno_graph analyze for Worker constructors, remove this flag, then have a requirement for `deno compile` that everything needs to be statically analyzable for now?

aapoalas replied:
While Worker constructs are not 100% statically analysable due to the overriding issue, I do have to admit that it's probably a good idea. Overriding the constructor is anyways a very minor case, even less so with a `compile` usage on top.

If we leave it at that, supporting statically analysable dynamic imports and Workers without providing this flag, then the end result will be that users with more complex `import() + compile` or `Worker + compile` cases will be forced to use my hacky idea of eg. including arrays of `() => import("./static-name.ts")` functions in their code to provide those modules into the module graph.

That is not really the worst situation imaginable. I could live with that. But, I also am a great fan of hacky solutions that skirt the edge between what is possible with the tools given and what is proper.

nayeemrmn replied:
> The real answer would be the [asset reference proposal](https://github.com/tc39/proposal-asset-references) though even then the question of how to do types is a bit tough.

At this point the more relevant proposals for this are [module expressions](https://github.com/tc39/proposal-module-expressions), [module declarations](https://github.com/tc39/proposal-module-declarations) and [import reflection](https://github.com/tc39/proposal-import-reflection). Since those are actually progressing, there is a real concern of future-proofness with starting to statically analyse `Worker` specifiers.

Is it out of the question to just... allow compiled modules to use the runtime module loader and file system cache on the executing machine, and print warnings when it ever needs to do so / error if incapable? It will always be a subpar experience when something works with `deno run` but not with `deno compile`. I don't know if it's a goal to try (again) to use a stripped down executable with compiled scripts but that seems like a maintenance burden anyway.

aapoalas replied:
@dsherret 's original point has gotten somewhat lost in this thread but was very much correct: `.help("Some help here")` should be added to explain the option.

andreubotella replied:
Done. Hopefully the long help is descriptive enough to the average user.

dsherret commented:
@aapoalas that one is #17405

andreubotella commented:
The `dynamic_import` test from this PR was split into #18017, and I've now updated the PR and added a new version of that test for dynamic imports that can't be statically analyzed. This is tested by reading the import path from the filesystem.

bartlomieju commented:
Discussed during CLI team meeting, we are in favor of landing this PR and going with a flag appears to be a better solution - it's easier to explain and document than requiring statically analyzable dynamic imports. I'm sure we'll bike shed on the flag name a bit. Additional benefit is that it gives a clear way to move this option to the configuration file in the future.

andreubotella commented:
Rebased to not depend on #17657 anymore.

aapoalas commented on cli/tests/testdata/compile/dynamic_imports/import_path:
**nitpick:** Missing newline.

aapoalas reviewed (changes requested):
Should add a help comment to the command line argument, otherwise looks ready for merge.

bartlomieju commented on cli/args/flags.rs:
```diff
@@ -906,6 +907,20 @@ fn compile_subcommand<'a>() -> Command<'a> {
+        .long("side-module")
+        .help("Additional module to include in the module graph")
+        .long_help(
+          "Includes an additional module in the compiled executable's module \
+    graph. Use this flag if a dynamically imported module or a web worker main \
+    module fails to load in the executable. This flag can be passed multiple \
+    times, to include multiple additional modules.",
```
Can I do comma separated list of urls like: `--side-module=https://deno.land/std/flags/mod.ts,https://deno.land/std/version.ts`?

andreubotella replied:
Commas are valid in URLs and filenames.

bartlomieju replied:
Hm, I thought we already supported comma separated lists in other flags, but maybe I'm wrong here

bartlomieju commented on cli/graph_util.rs:
```diff
@@ -150,7 +150,7 @@ pub fn graph_lock_or_exit(graph: &ModuleGraph, lockfile: &mut Lockfile) {
 }
 
 pub async fn create_graph_and_maybe_check(
```
I think we should leave this method as is and add additional `create_graph_with_side_modules_and_maybe_check`. @dsherret thoughts on this one? I remember you changed this API recently to have a single root module.

bartlomieju replied:
@dsherret any thoughts here?

dsherret replied:
I feel like this is fine and preferred? What's the reason to have a separate function to restrict it to a single root module here?

bartlomieju commented on cli/args/flags.rs:
```diff
@@ -906,6 +907,20 @@ fn compile_subcommand<'a>() -> Command<'a> {
+    .arg(
+      Arg::new("side-module")
+        .long("side-module")
+        .help("Additional module to include in the module graph")
```
Let's add a note that it's a new an unstable function
```suggestion
        .help("UNSTABLE: Additional module to include in the module graph")
```

bartlomieju reviewed (commented):
I tested it locally with a bunch of dynamic imports and while having to specify `--side-module=` multiple times is not super nice it does the job very well. I'm sure this functionality will unblock many new use cases for users and I'm strongly in favor of landing it. We can try to further iterate on the flag name and potentially making it more ergonomic, but it seems worth landing as is already.

Awesome work as always @andreubotella, LGTM.

bartlomieju reviewed (approved):
LGTM

dsherret commented on cli/tests/testdata/compile/dynamic_imports/main_unanalyzable.ts:
```diff
@@ -0,0 +1,18 @@
+  console.log("Dynamic importing");
+  const importPath = (await Deno.readTextFile(IMPORT_PATH_FILE_PATH)).trim();
+  import(importPath).then(() => console.log("Dynamic import done."));
+}, 500);
```
Why have a set timeout? Can we lower this so this test runs faster?

andreubotella replied:
This was for symmetry with `main.ts` (added in #18017), but there isn't actually a reason. I guess making it a zero timeout would be best. Should I change both in this PR, or in a follow-up?

dsherret replied:
Just changing both in this PR would be fine so we don’t forget. Zero timeout sounds good. Thanks!

andreubotella replied:
Done.

dsherret reviewed (approved):
LGTM too once the timeout is lowered. Thanks for contributing this and the other worker PR! Will be nice to have.

nayeemrmn commented on cli/args/flags.rs:
```diff
@@ -908,6 +909,20 @@ fn compile_subcommand<'a>() -> Command<'a> {
     .trailing_var_arg(true)
     .arg(script_arg().required(true))
+    .arg(
+      Arg::new("side-module")
```
Was flag name settled? Me and @aapoalas suggested `--import` which is less colloquial than `--side-module`.

bartlomieju replied:
No - it's still up to debate/bike-shedding, but I'm fine changing the name in a follow up. I'm not sure I like `--import` better than `--side-module`, seems more ambigious.

ry replied:
`--include` ? Shorter is better here...

bartlomieju replied:
I'm fine with `--include` too

nayeemrmn replied:
Yeah `--include` was our next one https://github.com/denoland/deno/pull/17663#issuecomment-1427085167, sounds perfect if it might include assets as well.

ry replied:
ok let's switch to `--include`

bartlomieju commented:
> Tangential: the flag could potentially be reused to define file system assets (ffi dylib / game assets / worker source) to be included in the compiled binary? It is a very useful feature to have IMO.

I'm not sure `eszip` is able to handle these kinds of files, but if it can, we should explore this possibility.

> In any case, if you agree that the flag name should be --include, I'm fine with that.

@andreubotella let's go with `--include`.

andreubotella commented:
Flag renamed.

bartlomieju reviewed (approved):
LGTM, thank you @andreubotella!
